internal/mesher/controller_test.go

```diff
@@ -0,0 +1,217 @@
+	tests := []struct {
+		name            string
+		createResources func(*testing.T, *fake.Clientset)
+		updateResources func(*testing.T, *fake.Clientset)
+
+		wantService             *corev1.Service
+		wantMeshServiceDeletion bool
+	}{
+		{
+			name: "creates mesh service on add",
+			createResources: func(t *testing.T, c *fake.Clientset) {
+				t.Helper()
+				_, err := c.CoreV1().Services("app").Create(generateAppService(8080))
+				require.NoError(t, err)
+			},
+			wantService: generateMeshService(8080),
+		},
+		{
+			name: "update mesh service on add if it already exists",
+			createResources: func(t *testing.T, c *fake.Clientset) {
+				t.Helper()
+				_, err := c.CoreV1().Services("app").Create(generateAppService(8080))
+				require.NoError(t, err)
+
+				_, err = c.CoreV1().Services(meshNamespace).Create(generateMeshService(9090))
+				require.NoError(t, err)
+			},
+			wantService: generateMeshService(8080),
+		},
+		{
+			name: "updates mesh service",
+			createResources: func(t *testing.T, c *fake.Clientset) {
+				t.Helper()
+				_, err := c.CoreV1().Services("app").Create(generateAppService(9999))
+				require.NoError(t, err)
+			},
+			updateResources: func(t *testing.T, c *fake.Clientset) {
+				_, err := c.CoreV1().Services("app").Update(generateAppService(8080))
+				require.NoError(t, err)
+			},
+			wantService: generateMeshService(8080),
+		},
+		{
+			name: "creates mesh service on update if deleted",
+			createResources: func(t *testing.T, c *fake.Clientset) {
+				t.Helper()
+				_, err := c.CoreV1().Services("app").Create(generateAppService(9999))
+				require.NoError(t, err)
+			},
+			updateResources: func(t *testing.T, c *fake.Clientset) {
+				t.Helper()
+				err := c.CoreV1().Services(meshNamespace).Delete("maesh-foo-app", nil)
+				require.NoError(t, err)
+
+				_, err = c.CoreV1().Services("app").Update(generateAppService(8080))
+				require.NoError(t, err)
+			},
+			wantService: generateMeshService(8080),
+		},
+		{
+			name: "deletes mesh service",
+			createResources: func(t *testing.T, c *fake.Clientset) {
+				t.Helper()
+				_, err := c.CoreV1().Services("app").Create(generateAppService(8080))
+				require.NoError(t, err)
+			},
+			updateResources: func(t *testing.T, c *fake.Clientset) {
+				t.Helper()
+				err := c.CoreV1().Services("app").Delete("foo", nil)
+				require.NoError(t, err)
+			},
+			wantMeshServiceDeletion: true,
+			wantService:             generateMeshService(8080),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := context.Background()
+
+			clientSet := fake.NewSimpleClientset()
+			factory := informers.NewSharedInformerFactory(clientSet, 0)
+
+			controller := mesher.NewController(clientSet, factory, meshNamespace, logger())
+
+			factory.Start(ctx.Done())
+
+			test.createResources(t, clientSet)
+			factory.WaitForCacheSync(ctx.Done())
+
+			go controller.Run()
+
+			// Forces the processing of the createResources events by the controller.
```

Comment:
This is waiting for processing more than forcing processing, right?

Reply:
To actually force the scheduler to execute the informers goroutines instead of this one. But I don't really like that and I want to improve it.